File: src/game/protocol/projection.ts
// M3-4 View 动画投影(TS 端口,ADR-020)。Python harness/view/projection.py 的镜像。
// 一致性由 src/game/protocol/projection.test.ts 锁(对同一 fixture parity)。
import type { TerraEvent } from "../../ipc/subscribe";

export type NpcKind = "guide" | "verifier" | "reviewer" | "builder";
export type NpcState =
  | "idle" | "decomposing" | "thinking" | "working" | "rework"
  | "awaiting_review" | "verifying" | "reviewing" | "hitl" | "error";
export type Zone =
  | "yard" | "lobby" | "workshop" | "review_door"
  | "verify_room" | "review_room" | "at_glass";

export interface NpcSnapshot {
  kind: NpcKind;
  sprite_key: string;
  state: NpcState;
  zone: Zone;
  task_id: string | null;
  think: { text: string; event_id: number } | null;
}
export type TaskStatus =
  | "queued" | "building" | "awaiting_review" | "verifying" | "post_verify"
  | "reviewing" | "merged" | "blocked" | "rejected";
export interface TaskBoardEntry {
  status: TaskStatus;
  builder: string | null;
  last_event_id: number;
}
export interface ViewSnapshot {
  npcs: Record<string, NpcSnapshot>;
  task_board: Record<string, TaskBoardEntry>;
  last_merge: { task_id: string; event_id: number } | null;
  cursor: number | null;
}

const INSTANCE_RE = /^([a-z_][a-z0-9_]*)#[0-9]+$/;
const VERIFIER_ROLE = "blaster";
const REVIEWER_ROLES = new Set(["tailor", "appsec"]);
const FIXED = ["guide", "blaster#1", "tailor#1", "appsec#1"];

function roleOf(npcId: string): string {
  if (npcId === "guide") return "guide";
  const m = INSTANCE_RE.exec(npcId);
  return m ? m[1] : npcId;
}
function kindOf(npcId: string): NpcKind {
  const r = roleOf(npcId);
  if (r === "guide") return "guide";
  if (r === VERIFIER_ROLE) return "verifier";
  if (REVIEWER_ROLES.has(r)) return "reviewer";
  return "builder";
}
function spriteOf(npcId: string): string { return roleOf(npcId); }

function zoneOf(npcId: string, state: NpcState): Zone {
  const kind = kindOf(npcId);
  if (state === "hitl") return "at_glass";
  if (kind === "guide") return "lobby";
  if (state === "idle") {
    return kind === "verifier" ? "verify_room" : kind === "reviewer" ? "review_room" : "yard";
  }
  if (state === "thinking") {
    if (kind === "builder") return "workshop";
    return kind === "verifier" ? "verify_room" : kind === "reviewer" ? "review_room" : "lobby";
  }
  // 其余 state 直接由 state 决定
  const map: Partial<Record<NpcState, Zone>> = {
    working: "workshop", rework: "workshop", decomposing: "lobby",
    awaiting_review: "review_door", verifying: "verify_room", reviewing: "review_room",
  };
  return map[state]!;
}

// ── 显示用:角色的功能标签(干什么的;纯展示,不影响投影/parity)──────────
// 显示职能而非花名/昵称:如 tailor=代码审查、blaster=验证者。
export const ROLE_LABEL: Record<string, string> = {
  guide: "向导", frontend: "前端开发", backend: "后端开发", database: "数据库",
  desktop_shell: "桌面端", ai_engineer: "AI 工程", rapid_proto: "快速原型",
  tech_writer: "技术文档", mobile: "移动端", merchant: "通用开发",
  blaster: "验证者", tailor: "代码审查", appsec: "安全审查",
};
/** 实例 id(如 tailor#2)→ 功能标签(多实例加序号):"代码审查2";取不到则原样。 */
export function agentName(id: string): string {
  const label = ROLE_LABEL[roleOf(id)];
  if (!label) return id;
  const n = id.includes("#") ? id.split("#")[1] : "";
  return n && n !== "1" ? `${label}${n}` : label;
}

export function project(events: TerraEvent[]): ViewSnapshot {
  const npcs: Record<string, NpcSnapshot> = {};
  const delegateTask: Record<number, string> = {};      // guide_delegate.event_id → task_id
  const builderOfTask: Record<string, string> = {};     // task_id → builder 实例
  const taskBoard: Record<string, TaskBoardEntry> = {};
  let lastMerge: ViewSnapshot["last_merge"] = null;
  let cursor: number | null = null;

  const present = (id: string) => {
    if (!npcs[id]) {
      npcs[id] = { kind: kindOf(id), sprite_key: spriteOf(id),
                   state: "idle", zone: zoneOf(id, "idle"),
                   task_id: null, think: null };
    }
  };
  const touchTask = (tid: string | null | undefined, eid: number, status: TaskStatus,
                     builder?: string) => {
    if (!tid) return;
    const slot = taskBoard[tid] ?? { status, builder: builder ?? null, last_event_id: eid };
    slot.status = status;
    slot.last_event_id = eid;
    if (builder) slot.builder = builder;
    taskBoard[tid] = slot;
  };
  const setState = (id: string, state: NpcState, taskId: string | null = null) => {
    present(id);
    const cur = npcs[id];
    cur.zone = state === "error" ? cur.zone : zoneOf(id, state); // error 保留原区
    cur.state = state;
    if (taskId !== null) cur.task_id = taskId;
  };

  for (const n of FIXED) present(n); // 固定班子恒在场

  for (const e of events) {
    const eid = e.event_id;
    cursor = eid ?? cursor;
    const t = e.type;
    const p = (e.payload ?? {}) as Record<string, unknown>;
    const agent = e.agent;
    const tid = (p.task_id as string | undefined) ?? null;

    if (t === "user_command") setState("guide", "decomposing");
    else if (t === "guide_think") {
      setState("guide", "thinking");
      npcs["guide"].think = { text: (p.text as string) ?? "", event_id: eid };
    }
    else if (t === "guide_delegate") {
      const tc = p.task_card as { task_id: string } | undefined;
      if (tc) { delegateTask[eid] = tc.task_id; touchTask(tc.task_id, eid, "queued"); }
    }
    else if (t === "guide_assign") {
      const inst = p.assignee_instance as string;
      const dtid = delegateTask[p.task_card_event_id as number] ?? null;
      const k = kindOf(inst);
      if (k === "builder") {
        // 同一 builder 实例再次派 = 返工(首次出现 → working)
        setState(inst, inst in npcs ? "rework" : "working", dtid);
        if (dtid !== null) builderOfTask[dtid] = inst;
        touchTask(dtid, eid, "building", inst);
      } else if (k === "verifier") { setState(inst, "verifying", dtid); touchTask(dtid, eid, "verifying"); }
      else if (k === "reviewer") { setState(inst, "reviewing", dtid); touchTask(dtid, eid, "reviewing"); }
      setState("guide", "idle"); // 派完回大堂
    }
    else if (t === "npc_think") {
      setState(agent, "thinking");
      present(agent); npcs[agent].think = { text: (p.text as string) ?? "", event_id: eid };
    }
    else if (t === "tool_intent" || t === "tool_done") {
      setState(agent, "working"); touchTask(tid, eid, "building");
    }
    else if (t === "review_request") {
      setState(agent, "awaiting_review", tid); touchTask(tid, eid, "awaiting_review");
    }
    else if (t === "verify_run") {
      setState(agent, "idle"); touchTask(tid, eid, "post_verify");
    }
    else if (t === "review_verdict") {
      setState(agent, "idle");
      touchTask(tid, eid, p.verdict === "reject" ? "rejected" : "reviewing");
    }
    else if (t === "merge") {
      const mtid = p.task_id as string;
      const bi = builderOfTask[mtid];
      if (bi) setState(bi, "idle", mtid);
      if (p.result === "success") {
        lastMerge = { task_id: mtid, event_id: eid };
        touchTask(mtid, eid, "merged");
      } else touchTask(mtid, eid, "blocked"); // conflict
    }
    else if (t === "hitl_request") {
      setState("guide", "hitl", tid); touchTask(tid, eid, "blocked");
    }
    else if (t === "error" && agent) setState(agent, "error");
  }

  return { npcs, task_board: taskBoard, last_merge: lastMerge, cursor };
}

File: harness/view/projection.py
"""M3-1 View 动画投影(ADR-020)。

纯函数 reducer:顺序消费 Session 事件流,产出每个 NPC 的当前动画状态快照。
等价 Catch-up 快进(replay 到游标即得当前状态,ADR-001)。无副作用、不调 LLM。
状态/区/sprite-key 见 docs/contracts/animation_protocol.json + ADR-020。
"""
import json
import re
from pathlib import Path

_PROTO_FILE = Path(__file__).resolve().parents[2] / "docs" / "contracts" / "animation_protocol.json"
_PROTO = json.loads(_PROTO_FILE.read_text(encoding="utf-8"))
_STATES = set(_PROTO["states"])

_INSTANCE_RE = re.compile(r"^([a-z_][a-z0-9_]*)#[0-9]+$")
_VERIFIER_ROLE = "blaster"
_REVIEWER_ROLES = {"tailor", "appsec"}

# 固定班子(prompt-driven + 机械验证位):恒存在于小镇,缺省 idle 留守 home zone
_FIXED = ("guide", "blaster#1", "tailor#1", "appsec#1")


def _role_of(npc_id: str) -> str:
    if npc_id == "guide":
        return "guide"
    m = _INSTANCE_RE.match(npc_id)
    return m.group(1) if m else npc_id


def _kind(npc_id: str) -> str:
    role = _role_of(npc_id)
    if role == "guide":
        return "guide"
    if role == _VERIFIER_ROLE:
        return "verifier"
    if role in _REVIEWER_ROLES:
        return "reviewer"
    return "builder"


def _sprite_key(npc_id: str) -> str:
    return _role_of(npc_id)


def _zone(npc_id: str, state: str) -> str:
    """state(+ npc 类型)→ 空间区(ADR-020)。idle/thinking 随类型不同;error 由调用方保留原区。"""
    kind = _kind(npc_id)
    if state == "hitl":
        return "at_glass"
    if kind == "guide":
        return "lobby"  # 向导基本驻大堂(decomposing/thinking/idle/arbitrating)
    if state == "idle":
        return {"verifier": "verify_room", "reviewer": "review_room"}.get(kind, "yard")
    if state == "thinking":
        return "workshop" if kind == "builder" else \
            {"verifier": "verify_room", "reviewer": "review_room"}.get(kind, "lobby")
    return {  # 其余直接由 state 决定(builder 主体)
        "working": "workshop", "rework": "workshop", "decomposing": "lobby",
        "awaiting_review": "review_door", "verifying": "verify_room", "reviewing": "review_room",
    }[state]


def project(events: list[dict]) -> dict:
    """事件流 → View 动画快照。

    Returns:
        {
          "npcs": {npc_id: {"kind","sprite_key","state","zone","task_id"}},
          "last_merge": {"task_id","event_id"} | None,
          "cursor": last event_id | None,
        }
    """
    npcs: dict = {}
    delegate_task: dict = {}      # guide_delegate.event_id -> task_id
    builder_of_task: dict = {}    # task_id -> builder 实例
    last_merge = None
    cursor = None

    def present(npc_id: str):
        if npc_id not in npcs:
            npcs[npc_id] = {"kind": _kind(npc_id), "sprite_key": _sprite_key(npc_id),
                            "state": "idle", "zone": _zone(npc_id, "idle"), "task_id": None}

    def set_state(npc_id: str, state: str, task_id=None):
        present(npc_id)
        cur = npcs[npc_id]
        zone = cur["zone"] if state == "error" else _zone(npc_id, state)  # error 保留原区
        cur.update(state=state, zone=zone)
        if task_id is not None:
            cur["task_id"] = task_id

    for n in _FIXED:           # 固定班子恒在场
        present(n)

    for e in events:
        cursor = e.get("event_id", cursor)
        t = e["type"]
        p = e.get("payload", {})
        agent = e.get("agent")

        if t == "user_command":
            set_state("guide", "decomposing")
        elif t == "guide_think":
            set_state("guide", "thinking")
        elif t == "guide_delegate":
            delegate_task[e["event_id"]] = p["task_card"]["task_id"]
        elif t == "guide_assign":
            inst = p["assignee_instance"]
            tid = delegate_task.get(p.get("task_card_event_id"))
            kind = _kind(inst)
            if kind == "builder":
                # 同一 builder 实例再次派发 = 返工(首次出现 → working)
                set_state(inst, "rework" if inst in npcs else "working", tid)
            elif kind == "verifier":
                set_state(inst, "verifying", tid)
            elif kind == "reviewer":
                set_state(inst, "reviewing", tid)
            set_state("guide", "idle")  # 派完回大堂
        elif t == "npc_think":
            set_state(agent, "thinking")
        elif t in ("tool_intent", "tool_done"):
            set_state(agent, "working")
        elif t == "review_request":
            set_state(agent, "awaiting_review", p.get("task_id"))
        elif t == "verify_run":
            set_state(agent, "idle")            # 爆破工验完留守
        elif t == "review_verdict":
            set_state(agent, "idle")            # reviewer 审完留守
        elif t == "merge":
            tid = p.get("task_id")
            bi = builder_of_task.get(tid)
            if bi:
                set_state(bi, "idle", tid)      # 该卡 builder 回院子
            if p.get("result") == "success":
                last_merge = {"task_id": tid, "event_id": e["event_id"]}
        elif t == "hitl_request":
            set_state("guide", "hitl", p.get("task_id"))
        elif t == "error":
            if agent:
                set_state(agent, "error")

        # 维护 task -> builder 实例映射(供 merge 时回收)
        if t == "guide_assign":
            inst = p["assignee_instance"]
            if _kind(inst) == "builder":
                tid = delegate_task.get(p.get("task_card_event_id"))
                if tid is not None:
                    builder_of_task[tid] = inst

    return {"npcs": npcs, "last_merge": last_merge, "cursor": cursor}
